### PR TITLE
[TAS-66] Shard selection for insertion

### PR DESCRIPTION
**Pull request by AldoRastrelli**
# [TAS-66] [Subtask] Elección de Shard para Insertion

## Description
Shard election for insertion query, choosing the shard with the bigger available memory size.

### Changes
- Router keeps a shard_manager which keeps track of the available memory of each shard
- Router asks for MemoryUpdate when creating the health channel and after each time it has send an insertion to a shard
- ShardManager returns always the shard with the bigger available memory size.

- Router has been refactored. Small functions have been added to split bigger ones.
- Unused files have been deleted
- Cargo fmt has been run

### Keep in mind
- There might be a future need of router and shards communicating with other messages. The scope of this task does not include those interactions.

## Checklist
[Check all of the boxes you match, and explain why if not checked]
- [x] New unit tests added. Why (if not):
- [x] All unit tests passed

### CODE REVIEW

File: sharding/src/node/memory_manager.rs [hunk 1]
```diff
@@ -1,3 +1,4 @@
+use futures::future::ErrInto;
```
Comment (ncontinanza) [inline sharding/src/node/memory_manager.rs]:
```suggestion
```

File: sharding/src/node/router.rs [hunk 2]
```diff
@@ -1,92 +1,217 @@
 use postgres::{Client, NoTls};
 extern crate users;
+use crate::node;
```
Comment (ncontinanza) [inline sharding/src/node/router.rs]:
```suggestion
```

File: sharding/src/node/router.rs [hunk 3]
```diff
@@ -1,92 +1,217 @@
...
+use crate::node::messages::message::{Message, MessageType};
 use crate::utils::queries::query_is_insert;
 
 use super::node::*;
-use crate::utils::node_config::get_router_config;
+use super::shard_manager::ShardManager;
+use super::{node::*, shard};
```
Comment (ncontinanza) [inline sharding/src/node/router.rs]:
```suggestion
```

File: sharding/src/node/shard.rs [hunk 4]
```diff
@@ -1,31 +1,38 @@
+use futures::lock::Mutex;
 use inline_colorization::*;
 use postgres::{Client, NoTls};
+use std::borrow::BorrowMut;
```
Comment (ncontinanza) [inline sharding/src/node/shard.rs]:
```suggestion
```

File: sharding/src/node/shard.rs [hunk 5]
```diff
@@ -1,31 +1,38 @@
...
+use std::io::{Read, Write};
 use std::net::{TcpListener, TcpStream};
 use std::sync::{Arc, RwLock};
 use std::{io, thread};
 
 extern crate users;
 use super::memory_manager::MemoryManager;
+use super::messages::message::{Message, MessageType};
 use super::node::*;
+use crate::node::router::Channel;
```
Comment (ncontinanza) [inline sharding/src/node/shard.rs]:
```suggestion
```

File: sharding/src/node/shard.rs [hunk 6]
```diff
@@ -1,31 +1,38 @@
...
+use crate::node::{memory_manager, shard};
```
Comment (ncontinanza) [inline sharding/src/node/shard.rs]:
```suggestion
use crate::node::shard;
```

File: sharding/src/node/shard.rs [hunk 7]
```diff
@@ -1,31 +1,38 @@
...
 use crate::utils::node_config::get_shard_config;
 use rust_decimal::Decimal;
```
Comment (ncontinanza) [inline sharding/src/node/shard.rs]:
```suggestion
```

File: sharding/src/node/shard.rs [hunk 8]
```diff
@@ -75,46 +75,175 @@ impl<'a> Shard<'a> {
...
+            shard_clone.accept_connections(listener_clone, stream_clone);
+        });
+
+        shard
     }
 
     fn accept_connections(
+        &mut self,
         listener: Arc<RwLock<TcpListener>>,
         rw_stream: Arc<RwLock<Option<TcpStream>>>,
```
Comment (ncontinanza) [inline sharding/src/node/shard.rs:100]: This is now unused

File: sharding/src/node/shard.rs [hunk 9]
```diff
@@ -75,46 +75,175 @@ impl<'a> Shard<'a> {
...
+                    // Do nothing
+                }
+            };
+        }
+    }
+
+    fn get_response_message(&mut self, message: &str) -> Option<String> {
+        let message = match Message::from_string(&message) {
+            Ok(message) => message,
+            Err(e) => {
```
Comment (ncontinanza) [inline sharding/src/node/shard.rs]: `e` is unused if the print is commented
```suggestion
            Err(_e) => {
```

File: sharding/src/node/shard.rs [hunk 10]
```diff
@@ -75,46 +75,175 @@ impl<'a> Shard<'a> {
...
 
-impl<'a> NodeRole for Shard<'a> {
+impl NodeRole for Shard {
     fn send_query(&mut self, query: &str) -> bool {
-        let rows = match self.backend.query(query, &[]) {
-            Ok(rows) => {
-                println!("Query executed successfully: {:?}", rows);
-                rows
-            }
+        let rows = match self.backend.as_ref().try_lock().unwrap().query(query, &[]) {
```
Comment (ncontinanza) [inline sharding/src/node/shard.rs]:
```suggestion
        match self.backend.as_ref().try_lock().unwrap().query(query, &[]) {
```

**Review by ncontinanza — CHANGES_REQUESTED** [review]

File: sharding/src/node/shard.rs (hunk 8 above)
Comment (AldoRastrelli) [reply sharding/src/node/shard.rs:100]: yet

**Review by ncontinanza — APPROVED** [review]

**Pull request merged**